// @cloudbase/model/dist/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerModel = void 0;
var wx_cloud_client_sdk_1 = require("@cloudbase/wx-cloud-client-sdk");
var COMPONENT_NAME = 'models';
var getComponent = function (cloudbase) { return ({
    name: COMPONENT_NAME,
    namespace: COMPONENT_NAME,
    entity: ((0, wx_cloud_client_sdk_1.init)({
        callFunction: (cloudbase.callFunction).bind(cloudbase),
        auth: (cloudbase.auth).bind(cloudbase),
    })).models,
}); };
try {
    var component = getComponent(cloudbase);
    cloudbase.registerComponent(component);
}
catch (e) { }
function registerModel(app) {
    try {
        var component = getComponent(app);
        app.registerComponent(component);
    }
    catch (e) {
        console.warn(e);
    }
}
exports.registerModel = registerModel;
__exportStar(require("@cloudbase/wx-cloud-client-sdk"), exports);
