// @cloudbase/ai/src/models/HunYuan/type.ts
type HunYuanMessage = {
  Role: 'user' | 'system' | 'assistant' | 'tool'
  Content?: string
  Contents?: Array<{
    Type: 'text' | 'image_url'
    Text?: string
    ImageUrl?: string
  }>
  ToolCallId?: string
  ToolCalls?: Array<{
    Id: string
    Type: 'function'
    Function: {
      Name: string
      Arguments: string
    }
  }>
}

type HunYuanTool = {
  Type: 'function'
  Function: {
    Name: string
    Arguments: string
  }
}

export type HunYuanInputData = {
  Model:
  | 'hunyuan-lite'
  | 'hunyuan-standard'
  | 'hunyuan-standard-256K'
  | 'hunyuan-pro'
  | 'hunyuan-code'
  | 'hunyuan-role'
  | 'hunyuan-functioncall'
  | 'hunyuan-vision'
  Messages: Array<HunYuanMessage>
  StreamModeration?: boolean
  TopP?: number
  Temperature?: number
  EnableEnhancement?: boolean
  Tools?: Array<HunYuanTool>
  ToolChoice?: 'none' | 'auto' | 'custom'
  CustomTool?: HunYuanTool
  SearchInfo?: boolean
}

// todo: more precise
export type HunYaunGenerateTextOutput = {
  choices?: Array<{
    message?: { content?: string }
  }>
}

// todo: more precise
export type HunYaunStreamTextOutput = {
  choices?: Array<{
    delta?: { content?: string }
  }>
}
