// @cloudbase/ai/dist/cjs/utils.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createAsyncIterable = exports.TextDecoderStream = exports.toPolyfillReadable = exports.createEventSourceParserTransformStream = exports.TransformStream = exports.ReadableStream = void 0;
var web_streams_polyfill_1 = require("web-streams-polyfill");
var web_streams_adapter_1 = require("@mattiasbuelens/web-streams-adapter");
var eventsource_parser_1 = require("eventsource-parser");
var text_encoding_shim_1 = require("text-encoding-shim");
exports.ReadableStream = web_streams_polyfill_1.ReadableStream;
exports.TransformStream = web_streams_polyfill_1.TransformStream;
var createEventSourceParserTransformStream = function () {
    var parser;
    return new exports.TransformStream({
        start: function (controller) {
            parser = (0, eventsource_parser_1.createParser)(function (event) {
                if (event.type === 'event') {
                    controller.enqueue(event);
                }
            });
        },
        transform: function (chunk) {
            parser.feed(chunk);
        },
    });
};
exports.createEventSourceParserTransformStream = createEventSourceParserTransformStream;
exports.toPolyfillReadable = (0, web_streams_adapter_1.createReadableStreamWrapper)(exports.ReadableStream);
var TextDecoderStream = (function () {
    function TextDecoderStream(encoding, options) {
        if (encoding === void 0) { encoding = 'utf-8'; }
        if (options === void 0) { options = {}; }
        var _this = this;
        this.transform = new exports.TransformStream({
            transform: function (chunk, controller) {
                var value = _this.handle.decode(new Uint8Array(chunk), { stream: true });
                if (value) {
                    controller.enqueue(value);
                }
            },
            flush: function (controller) {
                var value = _this.handle.decode();
                if (value) {
                    controller.enqueue(value);
                }
                controller.terminate();
            },
        });
        this.handle = new text_encoding_shim_1.TextDecoder(encoding, options);
    }
    Object.defineProperty(TextDecoderStream.prototype, "encoding", {
        get: function () {
            return this.handle.encoding;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TextDecoderStream.prototype, "fatal", {
        get: function () {
            return this.handle.fatal;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TextDecoderStream.prototype, "ignoreBOM", {
        get: function () {
            return this.handle.ignoreBOM;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TextDecoderStream.prototype, "readable", {
        get: function () {
            return this.transform.readable;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TextDecoderStream.prototype, "writable", {
        get: function () {
            return this.transform.writable;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(TextDecoderStream.prototype, Symbol.toStringTag, {
        get: function () {
            return 'TextDecoderStream';
        },
        enumerable: false,
        configurable: true
    });
    return TextDecoderStream;
}());
exports.TextDecoderStream = TextDecoderStream;
function createAsyncIterable(stream) {
    var _stream = stream;
    _stream[Symbol.asyncIterator] = function () {
        var reader = stream.getReader();
        return {
            next: function () {
                return __awaiter(this, void 0, void 0, function () {
                    var _a, done, value;
                    return __generator(this, function (_b) {
                        switch (_b.label) {
                            case 0: return [4, reader.read()];
                            case 1:
                                _a = _b.sent(), done = _a.done, value = _a.value;
                                return [2, done ? { done: true, value: undefined } : { done: false, value: value }];
                        }
                    });
                });
            },
        };
    };
    return _stream;
}
exports.createAsyncIterable = createAsyncIterable;
