// @cloudbase/ai/dist/cjs/models/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZhiPuModel = exports.MoonshotModel = exports.YiModel = exports.DashScopeModel = exports.ArkModel = exports.HunYuanBetaModel = exports.HunYuanModel = exports.MODELS = void 0;
var HunYuan_1 = require("./HunYuan");
Object.defineProperty(exports, "HunYuanModel", { enumerable: true, get: function () { return HunYuan_1.HunYuanModel; } });
var Ark_1 = require("./Ark");
Object.defineProperty(exports, "ArkModel", { enumerable: true, get: function () { return Ark_1.ArkModel; } });
var DashScope_1 = require("./DashScope");
Object.defineProperty(exports, "DashScopeModel", { enumerable: true, get: function () { return DashScope_1.DashScopeModel; } });
var Yi_1 = require("./Yi");
Object.defineProperty(exports, "YiModel", { enumerable: true, get: function () { return Yi_1.YiModel; } });
var Moonshot_1 = require("./Moonshot");
Object.defineProperty(exports, "MoonshotModel", { enumerable: true, get: function () { return Moonshot_1.MoonshotModel; } });
var ZhiPu_1 = require("./ZhiPu");
Object.defineProperty(exports, "ZhiPuModel", { enumerable: true, get: function () { return ZhiPu_1.ZhiPuModel; } });
var HunYuanBeta_1 = require("./HunYuanBeta");
Object.defineProperty(exports, "HunYuanBetaModel", { enumerable: true, get: function () { return HunYuanBeta_1.HunYuanBetaModel; } });
exports.MODELS = {
    hunyuan: HunYuan_1.HunYuanModel,
    'hunyuan-beta': HunYuanBeta_1.HunYuanBetaModel,
    ark: Ark_1.ArkModel,
    dashscope: DashScope_1.DashScopeModel,
    '01-ai': Yi_1.YiModel,
    moonshot: Moonshot_1.MoonshotModel,
    zhipu: ZhiPu_1.ZhiPuModel,
};
