// @cloudbase/ai/src/models/Moonshot/index.ts
import { type ParsedEvent } from 'eventsource-parser'
import {
  createAsyncIterable,
  TransformStream,
  TextDecoderStream,
  createEventSourceParserTransformStream,
  toPolyfillReadable,
} from '../../utils'
import type { MoonshotInputData, MoonshotGenerateTextOutput, MoonshotStreamTextOutput } from './type'
import type { ChatModel, StreamTextResult, ModelReq } from '../../type'

export class MoonshotModel implements ChatModel {
  public subUrl = 'moonshot/v1/chat/completions'
  constructor(private req: ModelReq, public baseUrl: string, subUrl?: string) {
    if (subUrl != null) {
      this.subUrl = subUrl
    }
  }

  private get url() {
    return `${this.baseUrl}/${this.subUrl}`
  }

  async generateText(data: MoonshotInputData) {
    const res = (await this.req({
      url: this.url,
      data: {
        ...data,
        stream: false,
      },
      stream: false,
    })) as MoonshotGenerateTextOutput
    return {
      text: (res?.choices?.[0]?.message?.content as string) ?? '',
      rawResponse: res,
    }
  }

  async streamText(data: MoonshotInputData) {
    const stream = await this.req({
      url: this.url,
      data: {
        ...data,
        stream: true,
      },
      stream: true,
    })
    return new MoonshotModelStreamResult(stream)
  }
}

class MoonshotModelStreamResult implements StreamTextResult {
  private _eventSourceStream: ReadableStream<ParsedEvent>

  constructor(_stream: ReadableStream<Uint8Array>) {
    const stream = toPolyfillReadable(_stream) as typeof _stream
    this._eventSourceStream = stream
      .pipeThrough(new TextDecoderStream())
      .pipeThrough(createEventSourceParserTransformStream())
  }

  private get teeedStream() {
    const [s1, s2] = this._eventSourceStream.tee()
    this._eventSourceStream = s2
    return s1
  }

  get eventSourceStream() {
    return createAsyncIterable(this.teeedStream)
  }

  get dataStream() {
    return createAsyncIterable(this.eventSourceStream.pipeThrough(new TransformStream<ParsedEvent, MoonshotStreamTextOutput>({
      transform(chunk, controller) {
        try {
          const data = JSON.parse(chunk.data) as MoonshotStreamTextOutput
          controller.enqueue(data)
        } catch (e) {
          if (chunk.data !== '[DONE]') {
            console.warn('Error when transforming event source data to json', e)
          }
        }
      },
    }),),)
  }

  get textStream() {
    return createAsyncIterable(this.dataStream.pipeThrough(new TransformStream<MoonshotStreamTextOutput, string>({
      transform(chunk, controller) {
        controller.enqueue(chunk?.choices?.[0]?.delta?.content ?? '')
      },
    }),),)
  }
}
