// @cloudbase/ai/dist/esm/models/DashScope/index.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { createAsyncIterable, TransformStream, TextDecoderStream, createEventSourceParserTransformStream, toPolyfillReadable, } from '../../utils';
var DashScopeModel = (function () {
    function DashScopeModel(req, baseUrl, subUrl) {
        this.req = req;
        this.baseUrl = baseUrl;
        this.subUrl = 'dashscope/compatible-mode/v1/chat/completions';
        if (subUrl != null) {
            this.subUrl = subUrl;
        }
    }
    Object.defineProperty(DashScopeModel.prototype, "url", {
        get: function () {
            return "".concat(this.baseUrl, "/").concat(this.subUrl);
        },
        enumerable: false,
        configurable: true
    });
    DashScopeModel.prototype.generateText = function (data) {
        var _a, _b, _c, _d;
        return __awaiter(this, void 0, void 0, function () {
            var res;
            return __generator(this, function (_e) {
                switch (_e.label) {
                    case 0: return [4, this.req({
                            url: this.url,
                            data: __assign(__assign({}, data), { stream: false }),
                            stream: false,
                        })];
                    case 1:
                        res = (_e.sent());
                        return [2, {
                                text: (_d = (_c = (_b = (_a = res === null || res === void 0 ? void 0 : res.choices) === null || _a === void 0 ? void 0 : _a[0]) === null || _b === void 0 ? void 0 : _b.message) === null || _c === void 0 ? void 0 : _c.content) !== null && _d !== void 0 ? _d : '',
                                rawResponse: res,
                            }];
                }
            });
        });
    };
    DashScopeModel.prototype.streamText = function (data) {
        return __awaiter(this, void 0, void 0, function () {
            var stream;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.req({
                            url: this.url,
                            data: __assign(__assign({}, data), { stream: true }),
                            stream: true,
                        })];
                    case 1:
                        stream = _a.sent();
                        return [2, new DashScopeModelStreamResult(stream)];
                }
            });
        });
    };
    return DashScopeModel;
}());
export { DashScopeModel };
var DashScopeModelStreamResult = (function () {
    function DashScopeModelStreamResult(_stream) {
        var stream = toPolyfillReadable(_stream);
        this._eventSourceStream = stream
            .pipeThrough(new TextDecoderStream())
            .pipeThrough(createEventSourceParserTransformStream());
    }
    Object.defineProperty(DashScopeModelStreamResult.prototype, "teeedStream", {
        get: function () {
            var _a = this._eventSourceStream.tee(), s1 = _a[0], s2 = _a[1];
            this._eventSourceStream = s2;
            return s1;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DashScopeModelStreamResult.prototype, "eventSourceStream", {
        get: function () {
            return createAsyncIterable(this.teeedStream);
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DashScopeModelStreamResult.prototype, "dataStream", {
        get: function () {
            return createAsyncIterable(this.eventSourceStream.pipeThrough(new TransformStream({
                transform: function (chunk, controller) {
                    try {
                        var data = JSON.parse(chunk.data);
                        controller.enqueue(data);
                    }
                    catch (e) {
                        if (chunk.data !== '[DONE]') {
                            console.warn('Error when transforming event source data to json', e);
                        }
                    }
                },
            })));
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DashScopeModelStreamResult.prototype, "textStream", {
        get: function () {
            return createAsyncIterable(this.dataStream.pipeThrough(new TransformStream({
                transform: function (chunk, controller) {
                    var _a, _b, _c, _d;
                    controller.enqueue((_d = (_c = (_b = (_a = chunk === null || chunk === void 0 ? void 0 : chunk.choices) === null || _a === void 0 ? void 0 : _a[0]) === null || _b === void 0 ? void 0 : _b.delta) === null || _c === void 0 ? void 0 : _c.content) !== null && _d !== void 0 ? _d : '');
                },
            })));
        },
        enumerable: false,
        configurable: true
    });
    return DashScopeModelStreamResult;
}());
